design/topology_proposal.md

```diff
@@ -0,0 +1,141 @@
+# Topology aware scheduling proposal
+
+## Background
+
+There is often interest in making the scheduler aware of factors such as
+availability zones. This document proposes a generic way to customize scheduler
+behavior based on labels attached to nodes.
+
+## Approach
+
+Define a repeated field under `Placement` that the scheduler consults when it
+places tasks. These "placement preferences" would be listed in decreasing order
+of precedence, and would have higher precedence than the default scheduler
+logic.
+
+These placement preferences would be interpreted based on their types, but
+the initially supported "spread over" message would tell the scheduler to spread
+tasks evenly between nodes which have each distinct value of the referenced node
+or engine label.
+
+## Protobuf changes
+
+In the `Placement` message that exists in `TaskSpec`, add a repeated field
+called `Preferences`. (Note: I don't think `Preferences` is a very good name.
+Looking for suggestions.)
+
+```
+repeated PlacementPreference preferences = 2;
+```
+
+`PlacementPreference` will be a message that specifies how to act on a label.
+The initially supported preference would be "spread" and "pack".
+
+```
+message SpreadOver {
+	string descriptor = 1; // label descriptor, such as engine.labels.az
+	// TODO: support node information beyond engine and node labels
+
+	// TODO: in the future, add a map that provides weights for weighted
+	// spreading.
+}
+
+message PlacementPreference {
+	oneof Preference {
+                SpreadOver spread = 1;
+        }
+
+	Preference pref = 1;
+}
+```
+
+## Behavior
+
+A simple use of this feature would be to spread tasks evenly between multiple
+availability zones. The way to do this would be to create an engine label on
+each node indicating its availability zone, and then create a
+`PlacementPreference` with type `SpreadOver` which references the engine label.
+The scheduler would prioritize balance between the availability zones, and if
+it ever has a choice between multiple nodes in the preferred availability zone
+(or a tie between AZs), it would choose the node based on its built-in logic.
+As of 1.13, this logic will prefer to schedule a task on the node which has the
+fewest tasks associated with the particular service.
+
+A slightly more complicated use case involves hierarchical topology. Say there
+are two datacenters, which each have four rows, each row having 20 racks. To
+spread tasks evenly at each of these levels, there could be three `SpreadOver`
+messages in `Preferences`. The first would spread over datacenters, the second
```

Comment:
Internally, the scheduler will be sorting the nodes to rank them. Basically, these rules define a ranking function for the nodes.

Using the example above, the sort function first takes into account the datacenter label. If datacenter A has 19 tasks from this service and datacenter B has 20 tasks from this service, all nodes with the `datacenter=A` label will rank above the datacenter B nodes at that moment. If any rows in datacenter A are more lightly populated than other rows, the nodes in those rows will be the highest ranked. And If any racks in those rows are more lightly populated than the other racks in those rows, their nodes will be ranked at the very top.

Suppose we start in this situation and want to scale up by two tasks. The first task will go to a node in datacenter A, unless that's not possible because of constraints. After the first task is scheduled, our instantaneous rankings change. Now both datacenters have the same number of tasks from the service, so the rankings instead depend what's going on at the row level. If there is one row in either datacenter that has fewer tasks than the other, the nodes in that row will be ranked highest. And so on.